### PR TITLE
Thread-based API for communicating with RE Manager via ZMQ.

### PR DESCRIPTION
**Pull request by dmgav**
Implemented thread-based API for communicating with RE Manager from 3rd party applications. The QueueServer provided asyncio-based API (class `ZMQCommSendAsync`) used by HTTP Server and CLI tool. The thread-based version of the API (class `ZMQCommSendThreads`) proposed in this PR provides similar functionality for applications that are not running asyncio loop (e.g. PyQT applications).

The API allows to send ZMQ requests to RE Manager in blocking mode or using callbacks. 

Blocking mode:
```
    client = ZMQCommSendThreads()

    try:
        msg = client.send_message(
            method="queue_item_add", params={"plan": _plan1, "user":  "Some User", "user_group": "admin"}
        )
        print(f"Response is received from RE Manager: {msg}")
    except CommTimeoutError as ex:
        print(f"Communication error occurred: {ex}") 
```

The same functionality can be implemented using callbacks:
```
    client = ZMQCommSendThreads()

    def cb(msg, msg_err):
        # Some code for handling the received data
        if msg_err:        
            print(f"Response is received from RE Manager: {msg}")
        else:
            print(f"Communication error occurred: {msg_err}")

    resp1 = client.send_message(
        method="queue_item_add", 
        params={"plan": _plan1, "user": _user, "user_group": _user_group},
        cb=cb,
    )
```

This PR contains all commits from https://github.com/bluesky/bluesky-queueserver/pull/98

### CODE REVIEW

**Review by mrakitin — COMMENTED** [review]
Looks good to me. Thanks for the very detailed write-up and thorough tests. I spotted a few incorrections, mostly in docstrings.

File: bluesky_queueserver/manager/comms.py [hunk 1]
```diff
@@ -416,6 +416,341 @@ def _pipe_receive(self):
...
+            msg_received.append(msg)
+            msg_err_received.append(msg_err)
+
+        zmq_comm.send_message(method="some_method", params={"some_value": n}, cb=cb)
+
+        # Wait for the message.
+        while not msg_received:
+            time.sleep(0.01)
+
+        # Code to process received 'msg'
```
[the remark this answers is not in the document]
Comment (dmgav) [reply bluesky_queueserver/manager/comms.py:473]: Actually no. While-loop will be broker once  `msg_received` contains at least one item, which will happen once the response is received. So there should be no indentation here. This is code example in docstring, but I believe it is correct.

File: bluesky_queueserver/manager/comms.py [hunk 2]
```diff
@@ -416,6 +416,341 @@ def _pipe_receive(self):
...
+        """
+        return {"method": method, "params": params}
+
+    def send_message(self, *, method, params=None, cb=None, raise_exceptions=None):
+        """
+        Send message to ZMQ server and wait for the response. The message must contain
+        a name of a method supported by the server and a dictionary of parameters that
+        are required by the method. In case of communication error (timeout), the function
+        returns error message or raises ``CommTimeoutError`` exception depending on
+        the setting of ``raise_timeout_exceptions`` property.
```
[the remark this answers is not in the document]
Comment (dmgav) [reply bluesky_queueserver/manager/comms.py]: Done

**Comment by dmgav** [comment]
@mrakitin Thank you for detailed review. I accepted all suggestions except one.

File: bluesky_queueserver/manager/comms.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrakitin) [reply bluesky_queueserver/manager/comms.py:473]: Oh, right, it makes sense. I suggested it before looking into the test examples, which use that a lot.

**Review by mrakitin — APPROVED** [review]

**Comment by mrakitin** [comment]
Ready to go, once the tests pass, but there we only docstring changes, so no code changes were introduced. I think it's safe to merge now.

**Pull request merged**